feat(combo-button): add component with tests and story

jendowns opened the pull request:
## Affected issues

- Resolves #66

## Proposed changes

- add `ComboButton` and `ComboButtonItem` with tests and story

diego-codes commented on src/components/ComboButton/ComboButton.stories.js:
```diff
@@ -0,0 +1,46 @@
+    TooltipDirection.TOP
+  ),
+});
+storiesOf(patterns('ComboButton'), module)
```
What's the difference between a pattern and a component?

jendowns replied:
A pattern is like a "complex component" that is made up of several pieces. And then a component is more like something core (and usually proxied from Carbon).

@petervachon did you have anything extra to add to this explanation? ☝️

diego-codes commented on src/components/ComboButton/ComboButton.stories.js:
```diff
@@ -0,0 +1,46 @@
+  .add('default', () => <ComboButton {...props()} actions={actions} />)
+  .add('single item', () => <ComboButton {...props()} actions={[actions[0]]} />)
+  .add('overflow items', () => {
+    const newActions = [...actions];
```
I think the purpose of the overflow items story is to have lots of actions to demonstrate the max-height styles of the overflow menu.
```suggestion
    const newActions = [...actions, {
      label: 'Filter list',
      renderIcon: Filter20,
      onClick: action('Filter list 1'),
    },
    {
      label: 'Filter list',
      renderIcon: Filter20,
      onClick: action('Filter list 2'),
    },
    {
      label: 'Filter list',
      renderIcon: Filter20,
      onClick: action('Filter list 3'),
    }];
```

diego-codes commented on src/components/ComboButton/__tests__/ComboButton.spec.js:
```diff
@@ -0,0 +1,27 @@
+
+import actions from '../_mocks_';
+
+describe('ComboButton', () => {
```
It would be good to add tests to ensure that one action does not render the overflow menu

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+      <Button
+        {...firstAction}
+        renderIcon={restActions.length > 1 ? null : firstAction.renderIcon}
+        data-testid="combo-button-primary-button"
```
It doesn't seem like you're using testids in your tests so I would remove them.

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+};
+
+const ComboButton = props => {
+  const [overflowMenuIcon, setOverflowMenuIcon] = useState(
```
It might be better to toggle a boolean state value instead of a component object because I would think that objects would have to be serialized to be compared. You can then use the boolean value to determine which icon to render.

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+  const getMenuOffset = () => {
+    const { top } = wrapper.current.getBoundingClientRect();
+    return {
+      top: isTop ? top : top * -1,
```
It would be better to divide by `-1` instead of multiply because if `top` is `0` the result would be `-1` instead of `0`

```suggestion
      top: isTop ? top : top / -1,
```

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+    const { top } = wrapper.current.getBoundingClientRect();
+    return {
+      top: isTop ? top : top * -1,
+      left: 'auto',
```
Should the left value also be determined by the bounding client rect as well? I think `auto` might be relying on browser implementations of `auto` which may be different.

jendowns replied:
Good point but honestly I'm unsure what other value this should be 🤔

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+  return (
+    <div
+      className={classnames(`${namespace}__wrapper`, className)}
+      data-testid="combo-button"
```
It doesn't seem like you're using testids in your tests so I would remove them.

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+            `${prefix}--btn--primary`
+          )}
+          menuOptionsClass={`${prefix}--list-box__menu`}
+          data-testid="combo-button-overflow-button"
```
It doesn't seem like you're using testids in your tests so I would remove them.

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+
+  return (
+    <div
+      className={classnames(`${namespace}__wrapper`, className)}
```
I would think that we could get away with just using the namespace as the root element classname instead of `__wrapper`
```suggestion
      className={classnames(namespace, className)}
```

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+  ),
+
+  /** @type {array<string>} Array of actions. */
+  actions: PropTypes.arrayOf(
```
I'm rethinking the idea that accepting an array of options makes sense. Similar to the `OverflowMenu` that accepts `OverflowMenuItem`s as children, this component should probably do the same and accept  `ComboButtonItem`s or `ComboButtonAction`s as children.

I'm not sure how feasible this would be but it would be the ideal API for this component for debugging and consistency reasons. It would make debugging much easier than debugging object shapes in an array as a prop.


Example:
```jsx
<ComboButton>
  <ComboButtonItem renderIcon={Icon} onClick={() => {}}>
    Action 1
  </ComboButtonItem>
  <ComboButtonItem renderIcon={Icon} onClick={() => {}}>
    Action 2
  </ComboButtonItem>
</ComboButton>

diego-codes commented on src/components/ComboButton/_mixins.scss:
```diff
@@ -0,0 +1,95 @@
+    display: flex;
+    justify-content: flex-end;
+
+    & .#{$prefix}--btn--primary:first-child {
```
For styling purposes of this component, I would suggest that combo-button classes are better than repurposing component classnames that we don't have control over.

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+          className={classnames(
+            buttonNamespace,
+            `${prefix}--btn`,
+            `${prefix}--btn--primary`
```
You could add a combo button specific classname for styling purposes specifically for this component.

jendowns replied:
Just clarifying: understanding is that these specific classes are needed to apply some of the Carbon `Button` styling to the `OverflowMenu`

Definitely agree though that I should add a new specific security-namespaced class for custom styling 👍

diego-codes replied:
Yeah, I believe you're right and I'm with you on using our own namespace for custom styling ⭐️

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+            `${prefix}--btn`,
+            `${prefix}--btn--primary`
+          )}
+          menuOptionsClass={`${prefix}--list-box__menu`}
```
I'm pretty sure this the overflow menu component already provides this classname 🤔

jendowns replied:
I think this is the *one* class that isn't getting added. 🤔 Or at least when I tried to remove it, the class was not added (and that prevents a `maxHeight` from getting added so the menu scrolls)

diego-codes replied:
hmm that's bizarre

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+    const itemClassName = `${prefix}--list-box__menu-item`;
+    return items.map((item, index) => (
+      <OverflowMenuItem
+        className={itemClassName}
```
I'm pretty the overflow menu component already provides this classname also 🤔

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,137 @@
+            <span className={`${itemClassName}-label`} title={item.label}>
+              {item.label}
+            </span>
+            <item.renderIcon className={`${itemClassName}-icon`} />
```
Should the icon have an empty description?

jendowns replied:
@diego-codes I think it should be good since `iconDescription` no longer has a default here. So nothing extra gets added to the svg and it keeps `aria-hidden="true"`.

But I did go ahead and remove that `title` from the `span` because it doesn't seem necessary. 🤔 What do you think about it now?

diego-codes replied:
I think the `title` is necessary for when the text overflows out of the allowed space so it is truncated with an ellipsis.

jendowns replied:
oh oops, you are right -- will add that back 👍

jendowns replied:
Hey one thing about this: now that we're using `ComboButtonItem`, which is basically just a renamed `OverflowMenuItem` that allows users to pass in any old node via the `itemText` prop, they would likely need to control this truncation themselves 🤔 I added an example & notes to the storybook on how to do that though.

jendowns commented:
Hey @diego-codes this is ready to look at again! I refactored to follow the structure you suggested above: https://github.com/carbon-design-system/ibm-security/pull/68#discussion_r321819098

I also added a note about truncation and how it's going to be up to the user now: https://github.com/carbon-design-system/ibm-security/pull/68#discussion_r322415873

Let me know what you think! Thanks 👍

diego-codes commented on src/components/ComboButton/__tests__/ComboButton.spec.js:
```diff
@@ -0,0 +1,52 @@
+    );
+
+    expect(comboButton).toMatchSnapshot();
+    expect(comboButton.render()).toMatchSnapshot();
```
If you're already doing a snapshot assertion with the render method, do you really need the snapshot test above?

jendowns replied:
Good catch - I'm going to keep line 49 & remove line 48. 👍

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,113 @@
+
+ComboButton.propTypes = {
+  /** @type {Object.<string, string>} An object list of props for the primary button. */
+  primaryAction: PropTypes.shape({
```
I was thinking that the primary action would be gathered from the children group. For example usage like this:

```jsx
<ComboButton>
  <ComboButtonItem renderIcon={Icon} onClick={() => {}}>
    Action 1
  </ComboButtonItem>
  <ComboButtonItem renderIcon={Icon} onClick={() => {}}>
    Action 2
  </ComboButtonItem>
</ComboButton>
```
Would take the first action and make it the primary action, basically abstracting away that logic within the component. Alternatively, a consumer can tag one of the `ComboButtonItem`s as the primary action:

```jsx
<ComboButtonItem renderIcon={Icon} onClick={() => {}} primary>Action 1</ComboButtonItem>
```

What do you think? Is this something that is doable and makes sense? My main motivation is to focus on flat props (instead of objects) which are easier to work with and debug IMO.

jendowns replied:
Ah yeah that makes sense! 👍 Sounds good to me; I can make that change

jendowns replied:
Well @diego-codes I was thinking about this & it may be difficult to achieve 🤔 

Right now `ComboButtonItem` is just a renamed `OverflowMenuItem` (side note: doesn't *have* to be this way, but it was definitely the path of least resistance, otherwise I was getting strange focus / tabbing results that I messaged you in Slack about).

The props for `ComboButtonItem` are pretty different from the props for `Button` (which is the underlying "primary action"). I notice in your mocked example you have button props like `renderIcon` etc but those aren't available to `ComboButtonItem` currently.

To that end, I think the biggest challenge is the `ComboButtonItem` / `OverflowMenuItem` components' `itemText` prop. That prop accepts nodes / anything really, so the user has a lot of freedom there. In the story examples I tried explaining how best to use `itemText` to get a well-formatted `ComboButton` in general, but it would be strange to take what's passed into `itemText`, especially when we are talking about maybe `span` elements and icon components, and then fit that into a `Button`. 🤔 

Anyway what do you think about all of this? 
I'm thinking that the 2 routes to solve this, if we want to stick with this general structure, are:
1. Keep `primaryActions` object
2. Have something like `ComboButtonAction` that renames a `Button` for this purpose, so it can be used as a sibling of `ComboButtonItem` components? Or just have a `Button` be used directly? My one worry here is that the API may be too open / lenient, it may imply that it's okay to put any number of other children here. One way to get around may be to add some warning if something other than `ComboButtonItem` or `ComboButtonAction`/`Button` are added as children of `ComboButton`?

What do you think? 🤔 😅

diego-codes replied:
I definitely see what you mean. I like the idea of having a separate component for the primary action. It reminds me of the Carbon UI Shell and DataTable components where it allows developers compose components with the provided building blocks.

I'm also thinking that another approach could be that the `ComboButtonItem` component becomes a shell component that renders either a `Button` or `OverflowMenu`. Pardon my over directing, here is what I could see working:

```jsx
const ComboButtonItem = props => {
  const { primary, renderIcon: Icon, children, className, onClick, iconDescription } = props;
  const classNames = classnames(namespace, { [`${namespace}--primary`]: primary }, className);
  if (primary) {
    return (
      <Button
        className={classNames}
        renderIcon={Icon}
        onClick={onClick}
        iconDescription={iconDescription}
      >
        {children}
      </Button>
    );
  }

  return (
    <OverflowMenuItem
      className={classNames}
      itemText={
        <Fragment>
          <span title={children}>{children}</span>
          <Icon iconDescription={iconDescription} />
        </Fragment>
      }
    />
  );
};

ComboButtonItem.propTypes = {
  primary: PropTypes.bool,
  className: PropTypes.string,
  renderIcon: PropTypes.node,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
  iconDescription: PropTypes.string,
};
```

This is just a hunch that it would work, but I haven't tested it. Either approach I think would work.

jendowns replied:
Thanks @diego-codes! This definitely makes sense & I appreciate you sharing your approach. 

One challenge that I see is that the primary `Button` and the `OverflowMenuItem` components are not siblings within `children` of the `ComboButton` component. 🤔 The `Button` needs to be outside of the underlying `OverflowMenu`,  whereas the `OverflowMenuItem` components need to be inside... I'm just wondering how best to accomplish that without adding too much complexity.

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,133 @@
+  const childrenArray = React.Children.toArray(children);
+
+  // Save first child (e.g., primary action) to use as a `Button`:
+  const primaryAction = React.cloneElement(childrenArray[0]);
```
I wonder if you need to actually clone the element first or if you can just use `childrenArray[0]`?
```suggestion  
  const primaryActionWithProps = [childrenArray[0]].map(
```

diego-codes commented on src/components/ComboButton/ComboButton.js:
```diff
@@ -0,0 +1,133 @@
+    overflowItems = childrenArray.slice(1);
+
+    // Create `OverflowMenuItem` components:
+    overflowMenuItemWithProps = React.Children.toArray(overflowItems).map(
```
I don't think you need to run this through `toArray` again since `overflowItems` is already an array

diego-codes commented on src/components/ComboButton/ComboButton.stories.js:
```diff
@@ -0,0 +1,84 @@
+                  onClick={action(`onClick ("${text}")`)}
+                >
+                  <span title={text}>{text}</span>
+                  <Filter20 />
```
I would make the API for the `ComboButtonItem` consistent. I like the use of the `renderIcon` that is used for the primary action instead of having the user provide their own children markup for the overflow actions.

diego-codes commented on src/components/ComboButton/ComboButtonItem/ComboButtonItem.js:
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import { namespace } from '../ComboButton';
+
+const ComboButtonItem = props => {
```
It's probably best to document the necessary prop types for this component otherwise consumers won't know what they need.

diego-codes commented on src/components/ComboButton/ComboButtonItem/ComboButtonItem.js:
```diff
@@ -0,0 +1,26 @@
+};
+
+ComboButtonItem.defaultProps = {
+  ...Button.defaultProps,
```
These props might be overwhelmingly long and conflicting possibly. It opens up way too much for a consumer to make sense of. For example `isDelete` does not apply to Buttons, but the prop types would say that it is a valid prop. It would force consumers to worry too much about the implementation of the `ComboButton`. I would just list out the specific ones that you need to render and nothing else.

They could just be: 
```js
{
  children: PropTypes.string.isRequired,
  renderIcon: PropTypes.node,
  onClick: PropTypes.onClick,
  disabled: PropTypes.bool,
  className: PropTypes.string,
}
```

Any other prop would just be confusing and unnecessary.

SimonFinney commented:
:tada: This PR is included in version 1.2.0-prerelease.5 :tada:

The release is available on:
- [npm package (@canary dist-tag)](https://www.npmjs.com/package/@carbon/ibm-security)
- [GitHub release](https://github.com/carbon-design-system/ibm-security/releases/tag/v1.2.0-prerelease.5%40canary)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

SimonFinney commented:
:tada: This PR is included in version 1.2.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@carbon/ibm-security)
- [GitHub release](https://github.com/carbon-design-system/ibm-security/releases/tag/v1.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
